Improve the version handling

williamdes opened the pull request:
This PR is to improve the Version handling and what goes with it.

mauriciofauth commented on scripts/create-release.sh:
```diff
@@ -120,24 +120,33 @@ mark_as_release() {
+VERSION_FILE=libraries/classes/Version.php
+
+fetchReleaseFromFile() {
+    php -r "require_once('libraries/classes/Version.php'); echo \PhpMyAdmin\Version::phpMyAdminVersion();"
```
I think `VERSION_SUFFIX` will never be defined here.

williamdes replied:
Indeed, It is only for packaging people to not do this https://salsa.debian.org/phpmyadmin-team/phpmyadmin/-/blob/debian/latest/debian/rules#L47 but change the vendor constant

mauriciofauth commented on libraries/classes/Version.php:
```diff
@@ -0,0 +1,62 @@
+        }
+
+        if (self::IS_PRE_RELEASE) {
+            return $versionRaw . '-' . self::PRE_RELEASE_NAME . self::PRE_RELEASE_NUMBER;
```
I think we could remove `IS_DEV` and `IS_PRE_RELEASE`, then add `dev` to `PRE_RELEASE_NAME` and only concatenate `PRE_RELEASE_NUMBER` if it is different than 0. Or drop `PRE_RELEASE_NUMBER` completely and add the number to  `PRE_RELEASE_NAME`. I think this will simplify the script.

mauriciofauth commented on libraries/classes/Version.php:
```diff
@@ -0,0 +1,69 @@
+    public const VERSION_MINOR = 1;
+    public const VERSION_PATCH = 0;
+
+    // The version will be {major}.{minor}.{patch}-dev
+    public const IS_DEV = true;
+
+    // The version will be {major}.{minor}.{patch}-{PRE_RELEASE_NAME}
+    public const IS_PRE_RELEASE = false;
```
I started to think that `IS_DEV` and `IS_PRE_RELEASE` will only add unnecessary steps to the Release Manager (as we can see in the `create-release.sh` script). If we need those boolean values, we can determine they based on `PRE_RELEASE_NAME` value.

williamdes replied:
I agree to in another hand this will make any simple code detection harder.
Let me know if you want me to proceed removing `IS_PRE_RELEASE` ?

But for `IS_DEV` it is too important to be removed (in a tools detection perspective)

My point is: people will build bad script parsing a version management they do not know, booleans are explicit ;p

ibennetch replied:
I would tend to agree with @mauriciofauth about this, but @williamdes has a valid point, too. I would leave IS_DEV in, but remove IS_PRE_RELEASE and refactor that check to see whether `PRE_RELEASE_NAME` is empty instead of also having the IS_PRE_RELEASE boolean. It's pretty minor in the scheme of things, so I won't insist on changing it.

mauriciofauth commented on libraries/classes/Version.php:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        if (defined('VERSION_SUFFIX')) {
+            return $versionRaw . VERSION_SUFFIX;
```
Is intended to only add `VERSION_SUFFIX` on released versions?

williamdes replied:
Yes, it is only for people packaging the final version ;)

mauriciofauth commented on libraries/classes/Version.php:
```diff
@@ -0,0 +1,69 @@
+    // The version will be {major}.{minor}.{patch}-{PRE_RELEASE_NAME}
+    public const IS_PRE_RELEASE = false;
+
+    public const PRE_RELEASE_NAME = '';
```
```suggestion
    public const PRE_RELEASE_NAME = 'rc1';
```

https://github.com/phpmyadmin/phpmyadmin/blob/ffbebaa8a27e646e90d658515e703da9fbb2596b/libraries/classes/Config.php#L139

williamdes replied:
This will have to be filled in when releasing, for now it should be empty

williamdes commented on libraries/classes/Routing.php:
```diff
@@ -25,7 +25,7 @@ public static function getDispatcher(): Dispatcher
         return cachedDispatcher($routes, [
             'cacheFile' => CACHE_DIR . 'routes.cache.php',
-            'cacheDisabled' => ($cfg['environment'] ?? '') === 'development',
+            'cacheDisabled' => ($cfg['environment'] ?? '') === 'development' || Version::isDev(),
```
Yes, devs, contributors, testers are not setting `environment` and fall into caching issues we have to handle afterwards on the GitHub issues.
It want to short circuit the problem: if DEV = routes can change = no cache

mauriciofauth replied:
We should improve the cache cleanup and warmup instead of disabling the cache for everyone. This way we can not enable cache in development environments.

williamdes replied:
For sure !
See https://github.com/nikic/FastRoute/issues/224 upstream

mauriciofauth replied:
I really don't like this change and the Template class as well. Having both sides of the Or operator doesn't make sense to me, because you should not set `$cfg['environment']` to `'development'` if you are not in a development version.

We should improve the documentation about the `$cfg['environment']` setting and improve the cache cleanup and warmup as I said before.

williamdes replied:
I understand and will revert the two changes
I agree we should try to fix this warm-up issue in another better way

ibennetch commented:
🤷‍♂️

I like the refactoring in general, but this seems to me to be a little complex. Still, it's an improvement, and I don't have any better idea, so go for it!

williamdes committed:
Improve the release script for versionning

Signed-off-by: William Desportes <williamdes@wdes.fr>

Improve the version handling

Signed-off-by: William Desportes <williamdes@wdes.fr>

Simplify the version handling

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes commented:
> I like the refactoring in general, but this seems to me to be a little complex. Still, it's an improvement, and I don't have any better idea, so go for it!

Thank you for the review !, I simplified slightly the code as you 2 suggested


@mauriciofauth do you agree that I merge this PR or do you have some more comments to do on it ?

mauriciofauth commented:
> do you agree that I merge this PR or do you have some more comments to do on it ?

@williamdes I just thought that we can have all those constants, but instead of manually update them, we can create a `scripts/console` command to auto-generate this class. This way we can have the version strings as constants as well.

ibennetch commented:
I was going to rework scripts/set-versions.sh to make this a bit less of a
manual process, but you’re welcome to jump on it too.

williamdes commented:
That seems like a good idea, not sure how I will do the automation but that will help for sure
Will work on that in some days more or less

mauriciofauth commented:
> I was going to rework scripts/set-versions.sh to make this a bit less of a manual process, but you’re welcome to jump on it too.

We could migrate this script to that new command as well.

> Will work on that in some days more or less

I can help you with that if you want.

williamdes commented:
> I can help you with that if you want.

Sure, you can push commits there or create a new PR

After a night of sleep I will have a day of driving so meanwhile you can do what seems best to you :) 

Just asking myself how we will manage the input of giving the version to the script to be parsed and re-built into constants

ibennetch commented:
> We could migrate this script to that new command as well.

I'm rather ignorant about the advantages of using scripts/console over the
shell script; is there something that makes that a better choice than a
shell script?

MauricioFauth committed:
Add the PhpMyAdmin\Version class generator

Creates the PhpMyAdmin\Command\SetVersionCommand class.

Usage: ./scripts/console set-version 5.1.0-dev

Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>

mauriciofauth commented:
> I'm rather ignorant about the advantages of using scripts/console over the shell script; is there something that makes that a better choice than a shell script?

Actually, we can add the `./scripts/console set-version 5.1.0-dev` line to the `set-version.sh` script and avoid having to migrate a shell script to PHP.

mauriciofauth commented:
I pushed the commit that adds the class generator. I renamed some constants to remove the redundant version word and I added the ID constant as well but I don't know if we'll going to use it or not.

williamdes commented on libraries/classes/Command/SetVersionCommand.php:
```diff
@@ -0,0 +1,105 @@
+
+    private function writeGeneratedClassFile(string $generatedClass): bool
+    {
+        $handle = fopen(ROOT_PATH . 'libraries/classes/Version.php', 'w');
+        if ($handle === false) {
+            return false;
+        }
+        fwrite($handle, $generatedClass);
+        fclose($handle);
+
```
why not using file_put_contents ?

mauriciofauth replied:
Good idea. For this case is a better idea to use it.

williamdes commented:
I like the changes you made
We need to:
- Have a lot of tests on the generation
- Correctness of the file, `php -l`
- Be sure to run the test suite **after** the version file has been generated

> I renamed some constants to remove the redundant version word and I added the ID constant as well but I don't know if we'll going to use it or not.

I think an important point is not if we will use it but if someone else will use it ;)

> I'm rather ignorant about the advantages of using scripts/console over the
shell script; is there something that makes that a better choice than a
shell script?

Nothing is better than a shell script and my point for that are:
- It never dies
- Is never deprecated
- Does not require any language to be setup
- Nobody does breaking changes to it (as far as I know)
- Does not require anybody to know a higher level of programming language and its specifics behaviors

mauriciofauth commented on scripts/create-release.sh:
```diff
@@ -120,24 +120,29 @@ mark_as_release() {
+VERSION_FILE=libraries/classes/Version.php
+
+fetchReleaseFromFile() {
+    php -r "require_once('libraries/classes/Version.php'); echo \PhpMyAdmin\Version::phpMyAdminVersion();"
```
```suggestion
    php -r "require_once('libraries/classes/Version.php'); echo \PhpMyAdmin\Version::VERSION;"
```

mauriciofauth commented on scripts/create-release.sh:
```diff
@@ -504,11 +511,20 @@ Todo now:
-        - in Documentation.html (if it exists) the 2 lines
-              " <title>phpMyAdmin 2.2.2-rc1 - Documentation</title> "
-              " <h1>phpMyAdmin 2.2.2-rc1 Documentation</h1> "
+        - in $VERSION_FILE Version class:
+            - for a dev cycle
+                - set IS_DEV to true
+                - check that PRE_RELEASE_NAME is empty
+            - for a normal release
+                - check that IS_DEV is false
+                - check that PRE_RELEASE_NAME is empty
+            - for a -rc release
+                - check that IS_DEV is false
+                - change PRE_RELEASE_NAME to "rc1"
```
This should be updated.

mauriciofauth commented on scripts/create-release.sh:
```diff
@@ -120,24 +120,29 @@ mark_as_release() {
 Please ensure you have incremented rc count or version in the repository :
-     - in $CONFIG_LIB Config::__constructor() the line
-          " \$this->set('PMA_VERSION', '$version'); "
+     - in $VERSION_FILE Version class:
+        - check that VERSION_MAJOR, VERSION_MINOR and VERSION_PATCH are correct.
+        - for a normal release
+            - check that IS_DEV is false
+            - check that PRE_RELEASE_NAME is empty
+        - for a -rc release
+            - check that IS_DEV is false
+            - change PRE_RELEASE_NAME to "rc1"
```
This should be updated.

williamdes replied:
Done

williamdes committed:
Add tests to generate the version file

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes committed:
Add tests for the Version class

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes committed:
Update the release script to match recent versionning changes

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes commented:
@ibennetch The work for the version handling is ready for a last review since you are the one that uses release scripts more
Will merge this as soon as you feel all this is okay :)

williamdes commented:
@ibennetch can I merge this one ?

williamdes committed:
Skip Console test for releases and packaging people

Signed-off-by: William Desportes <williamdes@wdes.fr>

mauriciofauth commented on libraries/classes/Command/SetVersionCommand.php:
```diff
@@ -0,0 +1,102 @@
+        $generatedClass = $this->getGeneratedClass($version);
+
+        if (! $this->writeGeneratedClassFile($generatedClass)) {
+            // failure
+            return 1;
```
```suggestion
            return Command::FAILURE;
```

mauriciofauth replied:
Forget about that, it will only work on `master`.

mauriciofauth commented on libraries/classes/Command/SetVersionCommand.php:
```diff
@@ -0,0 +1,102 @@
+
+        $output->writeln('PhpMyAdmin\Version class successfully generated!');
+
+        // success
+        return 0;
```
```suggestion
        return Command::SUCCESS;
```

mauriciofauth replied:
Forget about that, it will only work on `master`.

williamdes replied:
Good idea, I will add them on master
See: https://github.com/symfony/console/commit/27327a59875896d8b8e577ccddc35d92d7654664
it was added in console 5.1

williamdes replied:
66255ffbf91b1542eca8d64fa7b3cb5154bfd878
